### PR TITLE
Root directory per worker process

### PR DESCRIPTION
**Pull request by tagomoris**
This change introduces system-wide "root directory" and root directory per plugins.
It's enabled when:
* `root_dir` specified in `<system>` section
* `@id` configured in each plugin configuration sections

The root directories per plugin instances must be unique in a filesystem (between many workers of fluentd). So it uses `worker_id` provided by supervisor process via environment variable.
The path of directories are: `/system-wide-root/workerN/plugin-id` (generated and provided via `Fluent::Plugin::Base#plugin_root_dir`). Plugins will creates files or directories under that.

This change also adds 2 implementations to use this new feature:
* `buf_file`: configure buffer file `path` automatically
* `storage_local`: configure storage file `path` automatically

I have an additional plan to fix `in_tail` to use storage plugin as replacement of `pos_file`, but it's will be added by another pull-request.

### CODE REVIEW

**Comment by tagomoris** [comment]
@repeatedly could you review this change?

File: lib/fluent/plugin/buf_file.rb [hunk 1]
```diff
@@ -56,6 +51,14 @@ def initialize
       def configure(conf)
         super
 
+        unless @path
```
Comment (repeatedly) [inline lib/fluent/plugin/buf_file.rb]: In the future, should we validate `@path` value for avoding conflict trouble?
Fixed path should not allowed with multiprocess configuration or
support `%{worker_id}`/`%{plugin_id}` placeholder?

Comment (tagomoris) [inline lib/fluent/plugin/buf_file.rb]: Buffer chunks have `unique_id`, which are generated with time and random value.
So it SHOULD NOT conflict with other chunks in other processes. No need to do such thing.

Comment (tagomoris) [inline lib/fluent/plugin/buf_file.rb]: Ahhh, there are some problem about resuming at startup...

Comment (tagomoris) [inline lib/fluent/plugin/buf_file.rb]: Anyway, `plugin_id` can't help that situation. `plugin_id` is (should be) unique in a worker process, not server global.

File: lib/fluent/plugin_id.rb [hunk 2]
```diff
@@ -59,5 +64,17 @@ def plugin_id
         "object:#{object_id.to_s(16)}"
       end
     end
+
+    def plugin_root_dir
+      return @_plugin_root_dir if @_plugin_root_dir
+      return nil unless system_config.root_dir
```
Comment (repeatedly) [inline lib/fluent/plugin_id.rb]: Put `log.warn` is better?
Maybe, users hard to find the cause why `buffer_path`/`storage` is not configured automatically.

Comment (tagomoris) [inline lib/fluent/plugin_id.rb]: No, it's not about users.
This code is called by plugins, and plugin should check whether `plugin_root_dir` is available or not by return value, and raise errors with messages which shows how to configure that plugin.

**Comment by tagomoris** [comment]
Can I merge this?

**Comment by repeatedly** [comment]
LGTM

**Pull request merged**